### PR TITLE
Fix VTD/OSI smoketest

### LINKED ISSUE #18: VTD start script fails when symlinking the VTD setups
## Issue Description
Due to the heavy use of links within VTD setups and implications in symlinks pointing to parent directories, 
the default way Cloe's VTD start script arranges the 'VTD facades' fails with:
`[cloe] Error configuring simulator vtd: timeout waiting for SCP message`

## Expected Behavior
Once Cloe is built including the VTD packages, VTD related smoke tests should succeed.

## Steps to Reproduce
This can be reproduced by building Cloe and running VTD tests like this:
```
WITH_VTD=1 make export package
bats tests/test_vtd.bats
```

### PR DESCRIPTION
**Pull request by tobifalk**
In order to make the [OSMP Dummy Sensor Model](https://github.com/OpenSimulationInterface/osi-sensor-model-packaging/tree/master/examples/OSMPDummySensor) available in a simulation with VTD, the following steps have been taken:
* Copy instead of link the VTD setup to the simulation runtime directory (resolves #18).
* The sensor model library paths are expected to be contained in the environment variable VTD_EXTERNAL_MODELS, which is used to link the model libraries to the VTD setup directory.
* The [open-simulation-interface](https://github.com/OpenSimulationInterface/open-simulation-interface) Conan recipe now supports building different OSI versions and supports using different protobuf (and protoc) versions. 
* A Conan recipe for building the dummy sensor model as a VTD/OSI plugin is now available in the `vendor` directory. Since it has protobuf v2.6.1 as a dependency, the respective recipe has been added as well.
* `vtd-osi-sensor` is now a dependency of the Cloe meta package, if `WITH_VTD=1`.

If the local VTD installation is not equipped  with the VTD/OSI plugin, the OSI smoketest is skipped.

---
This change is [Reviewable](https://reviewable.io/reviews/eclipse/cloe/22)

### CODE REVIEW

File: vendor/open-simulation-interface/Makefile [hunk 1]
```diff
@@ -1,6 +1,16 @@
...
 override PACKAGE_CHANNEL := cloe/stable
 override CLEAN_SOURCE_DIR := true
 
 include ../../Makefile.package
+
+export:
+	@(echo "${VTD_OSI_VERSION}" > VERSION)
+	conan export . ${PACKAGE_NAME}/${VTD_OSI_VERSION}@${VTD_OSI_CHANNEL}
+	@(echo "${PACKAGE_VERSION}" > VERSION)
+	conan export . ${PACKAGE_NAME}/${PACKAGE_VERSION}@${PACKAGE_CHANNEL}
```
Comment (clssn) [inline vendor/open-simulation-interface/Makefile:16]: No newline at end of file. I suggest you configure the *Insert Final Newline*, *Render Final Newline* and *Trim Trailing Whitespace* options in VSCode to have these things taken care of automatically.

File: vendor/vtd-osi-sensor/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+)
+
+# Installation -------------------------------------------------------
+include(GNUInstallDirs)
+install(TARGETS ${PROJECT_NAME}
+    LIBRARY
+        DESTINATION ${CMAKE_INSTALL_LIBDIR}
+    ARCHIVE
+        DESTINATION ${CMAKE_INSTALL_LIBDIR}
+)
```
Comment (clssn) [inline vendor/vtd-osi-sensor/CMakeLists.txt:53]: Same here, missing newline.

**Review by clssn — APPROVED** [review]
Two minor findings. You can as well fix them in your next branch. Apart from that, LGTM!

**Comment by clssn** [comment]
Ah, you need to rebase onto master anyway, to resolve the conflict in `Makefile.all` so you could as well quickly insert the two missing newlines on the way ;-)

File: vendor/protobuf-2.6.1/conanfile.py [hunk 3]
```diff
@@ -19,6 +19,10 @@ class ProtobufConan(ConanFile):
         _patch_v2p6p1,
         "CMakeLists.txt",
     ]
+    requires = [
```
Comment (clssn) [inline vendor/protobuf-2.6.1/conanfile.py:22]: Shouldn't this be `build_requires`? Thos look like tools just needed for the build of this particular package...

Comment (tobifalk) [reply vendor/protobuf-2.6.1/conanfile.py]: That's right, I fixed it.

File: vendor/open-simulation-interface/Makefile (hunk 1 above)
Comment (tobifalk) [reply vendor/open-simulation-interface/Makefile]: fixed

File: vendor/vtd-osi-sensor/CMakeLists.txt (hunk 2 above)
Comment (tobifalk) [reply vendor/vtd-osi-sensor/CMakeLists.txt]: fixed

**Comment by tobifalk** [comment]
Just updated the vtd/osi plugin version. That should be all for this PR.

**Review by cassava — CHANGES_REQUESTED** [review]
> Reviewed 21 of 21 files at r1, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/eclipse/cloe/22)* status: 18 of 22 files reviewed, 13 unresolved discussions (waiting on @cassava and @tobifalk)

___
*[conanfile.py, line 54 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr4UEWD8MJLUHy5xsF:-MUr4UEXBcXAwUPoRvAZ:b-80wxnj) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/conanfile.py#L54)):*
> ```Python
>         if self.options.with_vtd:
>             requires.append("cloe-plugin-vtd")
>             self.requires("vtd-osi-sensor/1.0.0@cloe/vtd2p2")
> ```

Do we have a concept for how we are using the `name/version@user/channel` annotations?
What's the rationale for using `vtd2p2`?
___
*[plugins/vtd/bin/vtd, line 203 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr51Tz5U6iVj0Mr_Lm:-MUr51Tz5U6iVj0Mr_Ln:b-ibf5ka) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/plugins/vtd/bin/vtd#L203)):*
> ```
>     vtd_project: str,
>     copy_project: bool,
>     vtd_external_models : str,
> ```

This looks like the wrong type. Shouldn't it be `List[str]`?
___
*[plugins/vtd/bin/vtd, line 258 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr5UqQ56fp2Bhmgvlb:-MUr5UqQ56fp2Bhmgvlc:b-nn66hr) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/plugins/vtd/bin/vtd#L258)):*
> ```
>         if not dstfile.is_file():
>             logging.debug("Create symlink: {} -> {}".format(dstfile, srcfile))
>             os.symlink(srcfile, dstfile)
> ```

This modifies (by adding symlinks) to whatever `Current` is pointing to. 
According to the lines above this change, the current setup might be a symlink or might be copied. If it's a symlink, we probably shouldn't do this, right?

Choices;
- Always copy setups
- Check here whether setup was copied or not before modifying
- ???
___
*[tests/bootstrap_vtd.json, line 17 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr6_aR5xNoFYYC6fdk:-MUr6_aR5xNoFYYC6fdl:b-py4sfl) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/tests/bootstrap_vtd.json#L17)):*
> ```JSON
>             "start",
>             "--vtd-setup", "${THIS_STACKFILE_DIR}/../plugins/vtd/contrib/setups/Cloe.noGUInoIG",
>             "--copy-setup",
> ```

@clssn Did you want this to be the default?
___
*[vendor/open-simulation-interface/CMakeLists.txt, line 4 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr98hM7-EyIA7dy3Hq:-MUr98hM7-EyIA7dy3Hr:b-ly0qta) ([raw file](https://github.com/eclipse/cloe/blob/bf69501576e4eae03520c34b19cf20d65ea6c106/vendor/open-simulation-interface/CMakeLists.txt#L4)):*
> ```CMake
> 
> project(open_simulation_interface
>     VERSION 3.2.0
> ```

Why are we putting this in a separate VERSION file?
___
*[vendor/open-simulation-interface/conanfile.py, line 16 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr7tEC72PwS1eHkpX7:-MUr7tEC72PwS1eHkpX8:bi61zxs) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/vendor/open-simulation-interface/conanfile.py#L16)):*
> ```Python
>         "shared": [True, False],
>         "fPIC": [True, False],
>         "protoc_from_protobuf": [True, False],
> ```

Why do we have this option now?
___
*[vendor/open-simulation-interface/conanfile.py, line 40 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr7y9Y7G-UQL_Ri8WN:-MUr7y9Y7G-UQL_Ri8WO:b5d3s4m) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/vendor/open-simulation-interface/conanfile.py#L40)):*
> ```Python
>     _cmake = None
> 
>     def _project_version(self):
> ```

In the rest of the "project", this method refers to the repository root VERSION. Using the same name and methods here but changing the path is confusing. Call it `_package_version` if this is necessary. Or merge this with `set_version`.
___
*[vendor/open-simulation-interface/Makefile, line 1 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr8j_VBX5qyQXiIhmI:-MUr8j_VBX5qyQXiIhmJ:b6sr4vz) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/vendor/open-simulation-interface/Makefile#L1)):*
> ```Makefile
> VTD_OSI_VERSION := 3.0.1
> ```

I'm confused. What does this version have to do here? What does VTD have to do here, in `vendor/open-simulation-interface`? We shouldn't have such a hard coupling, if possible.
___
*[vendor/open-simulation-interface/Makefile, line 13 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr8_sI2peyt44tKSKz:-MUr8_sI2peyt44tKSL-:b-on3gth) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/vendor/open-simulation-interface/Makefile#L13)):*
> ```Makefile
> 
> export:
> 	@(echo "${VTD_OSI_VERSION}" > VERSION)
> ```

Are we modifying source files during build?
___
*[vendor/open-simulation-interface/Makefile, line 16 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr8XeB9nH53aJFsEvL:-MUr8XeB9nH53aJFsEvM:bu5vaoz) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/vendor/open-simulation-interface/Makefile#L16)):*
> ```Makefile
> 	conan export . ${PACKAGE_NAME}/${VTD_OSI_VERSION}@${VTD_OSI_CHANNEL}
> 	@(echo "${PACKAGE_VERSION}" > VERSION)
> 	conan export . ${PACKAGE_NAME}/${PACKAGE_VERSION}@${PACKAGE_CHANNEL}
> ```

Missing EOL
___
*[vendor/protobuf-2.6.1/conanfile.py, line 9 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr9JPE0HkeEvQeFpLE:-MUr9JPE0HkeEvQeFpLF:b-1qtq13) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/vendor/protobuf-2.6.1/conanfile.py#L9)):*
> ```Python
>     version = "2.6.1"
>     default_user = "cloe"
>     default_channel = "vtd2p2"
> ```

Can we avoid this coupling to vtd somehow?
___
*[vendor/vtd-osi-sensor/CMakeLists.txt, line 14 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr9hWrDAxgBE0J5PnE:-MUr9hWrDAxgBE0J5PnF:bvl2phm) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/vendor/vtd-osi-sensor/CMakeLists.txt#L14)):*
> ```CMake
> string(MD5 FMUGUID modelDescription.in.xml)
> configure_file(modelDescription.in.xml modelDescription.xml @ONLY)
> find_library(VTD_OSI_LIBRARY NAME libopen_simulation_interface.so
> ```

Let's be consistent with our formatting.
___
*[vendor/vtd-osi-sensor/CMakeLists.txt, line 52 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr9n4Y4eaRTLZWYE-M:-MUr9n4Y4eaRTLZWYE-N:bu5vaoz) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/vendor/vtd-osi-sensor/CMakeLists.txt#L52)):*
> ```CMake
>     ARCHIVE
>         DESTINATION ${CMAKE_INSTALL_LIBDIR}
> )
> ```

Missing EOL

**Review by cassava — CHANGES_REQUESTED** [review]
> Reviewed 3 of 4 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/eclipse/cloe/22)* status: all files reviewed, 13 unresolved discussions (waiting on @tobifalk)

___
*[Makefile.all, line 146 at r4](https://reviewable.io/reviews/eclipse/cloe/22#-MUrCHeA25tN3e7a6HEk:-MUrCHeA25tN3e7a6HEl:bb1v3td) ([raw file](https://github.com/eclipse/cloe/blob/f27afcf8908fe6d00aa7f3e0fc10b5b0fd778cc5/Makefile.all#L146)):*
> ```
> else
> ALL_PKGS := $(filter-out plugins/vtd, ${ALL_PKGS})
> ALL_VENDOR_PKGS := $(filter-out vendor/vtd*, ${ALL_PKGS})
> ```

Good idea! But I don't think you want to set ALL_VENDOR_PKGS to ALL_PKGS?

**Comment by clssn** [comment]
> > ALL_VENDOR_PKGS := $(filter-out vendor/vtd*, ${ALL_PKGS})
> Good idea! But I don't think you want to set ALL_VENDOR_PKGS to ALL_PKGS?

Ouch, my bad.

**Review by tobifalk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/eclipse/cloe/22)* status: 7 of 26 files reviewed, 11 unresolved discussions (waiting on @cassava and @clssn)

___
*[conanfile.py, line 54 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr4UEWD8MJLUHy5xsF:-MV11Be21ZOog5iR2jee:b-2swz2n) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/conanfile.py#L54)):*
<details><summary><i>Previously, cassava (Ben Morgan) wrote…</i></summary><blockquote>

Do we have a concept for how we are using the `name/version@user/channel` annotations?
What's the rationale for using `vtd2p2`?
</blockquote></details>

As discussed, we will use the `user/channel` to indicate the package origin (and quality). I appended`vtd2.2` to the package version instead.
___
*[plugins/vtd/bin/vtd, line 203 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr51Tz5U6iVj0Mr_Lm:-MV1LIhN7A4FJonR_t0C:bt2vr1b) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/plugins/vtd/bin/vtd#L203)):*
<details><summary><i>Previously, cassava (Ben Morgan) wrote…</i></summary><blockquote>

This looks like the wrong type. Shouldn't it be `List[str]`?
</blockquote></details>

Fixed.
___
*[plugins/vtd/bin/vtd, line 258 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr5UqQ56fp2Bhmgvlb:-MV1LMIH-gSyKYKBxFwe:b-xpv3x3) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/plugins/vtd/bin/vtd#L258)):*
<details><summary><i>Previously, cassava (Ben Morgan) wrote…</i></summary><blockquote>

This modifies (by adding symlinks) to whatever `Current` is pointing to. 
According to the lines above this change, the current setup might be a symlink or might be copied. If it's a symlink, we probably shouldn't do this, right?

Choices;
- Always copy setups
- Check here whether setup was copied or not before modifying
- ???
</blockquote></details>

Since we removed the option for sym-linking, no risk anymore.
___
*[tests/bootstrap_vtd.json, line 17 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr6_aR5xNoFYYC6fdk:-MV1LW-QA-Qd75ZSt8ax:b-hu20iz) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/tests/bootstrap_vtd.json#L17)):*
<details><summary><i>Previously, cassava (Ben Morgan) wrote…</i></summary><blockquote>

@clssn Did you want this to be the default?
</blockquote></details>

As discussed, vtd setups are now always copied.
___
*[vendor/open-simulation-interface/Makefile, line 1 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr8j_VBX5qyQXiIhmI:-MV1LeHs6rNarHErwui2:b-767cgm) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/vendor/open-simulation-interface/Makefile#L1)):*
<details><summary><i>Previously, cassava (Ben Morgan) wrote…</i></summary><blockquote>

I'm confused. What does this version have to do here? What does VTD have to do here, in `vendor/open-simulation-interface`? We shouldn't have such a hard coupling, if possible.
</blockquote></details>

True. I removed all vtd references in the osi package.
___
*[vendor/open-simulation-interface/Makefile, line 13 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr8_sI2peyt44tKSKz:-MV1Lna14Qis4utKsCuW:bf4u5za) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/vendor/open-simulation-interface/Makefile#L13)):*
<details><summary><i>Previously, cassava (Ben Morgan) wrote…</i></summary><blockquote>

Are we modifying source files during build?
</blockquote></details>

The osi package recipes for the different versions are now in two different directories, so no need anymore to determine the version from the Makefile case by case.
___
*[vendor/open-simulation-interface/Makefile, line 16 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr8XeB9nH53aJFsEvL:-MV1MVEF6FtwRVRMviPu:bt2vr1b) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/vendor/open-simulation-interface/Makefile#L16)):*
<details><summary><i>Previously, cassava (Ben Morgan) wrote…</i></summary><blockquote>

Missing EOL
</blockquote></details>

Fixed.
___
*[vendor/protobuf-2.6.1/conanfile.py, line 9 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr9JPE0HkeEvQeFpLE:-MV1NUF_AAy2nmUUeSdw:b-896fix) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/vendor/protobuf-2.6.1/conanfile.py#L9)):*
<details><summary><i>Previously, cassava (Ben Morgan) wrote…</i></summary><blockquote>

Can we avoid this coupling to vtd somehow?
</blockquote></details>

Done.
___
*[vendor/open-simulation-interface/CMakeLists.txt, line 4 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr98hM7-EyIA7dy3Hq:-MV1NAFUBbCqdXs5hKfA:bwx1xz9) ([raw file](https://github.com/eclipse/cloe/blob/bf69501576e4eae03520c34b19cf20d65ea6c106/vendor/open-simulation-interface/CMakeLists.txt#L4)):*
<details><summary><i>Previously, cassava (Ben Morgan) wrote…</i></summary><blockquote>

Why are we putting this in a separate VERSION file?
</blockquote></details>

VERSION file has been removed.
___
*[vendor/open-simulation-interface/conanfile.py, line 16 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr7tEC72PwS1eHkpX7:-MV1MYXHEqqjm6f4409T:bd2wcoh) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/vendor/open-simulation-interface/conanfile.py#L16)):*
<details><summary><i>Previously, cassava (Ben Morgan) wrote…</i></summary><blockquote>

Why do we have this option now?
</blockquote></details>

This option is needed for protobuf versions < 3.0 where there is no compatible proto compiler available on conan center. In those cases, we may use the protoc binary that comes with the protobuf library.
___
*[vendor/open-simulation-interface/conanfile.py, line 40 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr7y9Y7G-UQL_Ri8WN:-MV1MyIy0ARI1PN8_VDH:b-ahnyql) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/vendor/open-simulation-interface/conanfile.py#L40)):*
<details><summary><i>Previously, cassava (Ben Morgan) wrote…</i></summary><blockquote>

In the rest of the "project", this method refers to the repository root VERSION. Using the same name and methods here but changing the path is confusing. Call it `_package_version` if this is necessary. Or merge this with `set_version`.
</blockquote></details>

Has been removed during osi package refactoring (s. previous comment).
___
*[vendor/vtd-osi-sensor/CMakeLists.txt, line 14 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr9hWrDAxgBE0J5PnE:-MV1NH0wCY8y9pF3egcE:bt2vr1b) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/vendor/vtd-osi-sensor/CMakeLists.txt#L14)):*
<details><summary><i>Previously, cassava (Ben Morgan) wrote…</i></summary><blockquote>

Let's be consistent with our formatting.
</blockquote></details>

Fixed.
___
*[vendor/vtd-osi-sensor/CMakeLists.txt, line 52 at r1](https://reviewable.io/reviews/eclipse/cloe/22#-MUr9n4Y4eaRTLZWYE-M:-MV1NR_K78kuI_wwnmBk:bt2vr1b) ([raw file](https://github.com/eclipse/cloe/blob/5eeb0eb49702007fb9ab1d6f590268b8667e174d/vendor/vtd-osi-sensor/CMakeLists.txt#L52)):*
<details><summary><i>Previously, cassava (Ben Morgan) wrote…</i></summary><blockquote>

Missing EOL
</blockquote></details>

Fixed.

**Review by cassava — APPROVED** [review]
> Reviewed 1 of 1 files at r5, 10 of 14 files at r6, 1 of 1 files at r8, 1 of 1 files at r9, 8 of 8 files at r10.
> *[Reviewable](https://reviewable.io/reviews/eclipse/cloe/22)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @tobifalk)

**Pull request merged**